[MLIR][TableGen] Error on APInt parameter without custom comparator

Jezurko opened the pull request:
The error is triggered when an attribute or type uses an APInt typed parameter with the generated equality operator. If the compared APInts have different bit widths the equality operator triggers an assert. This is dangerous, since `StorageUniquer` for types and attributes uses the equality operator when a hash collision appears. As such, it is necessary to use custom provided comarator or `APIntParameter` that already has it.
This commit also replaces uses of the raw `APInt` parameter with the `APIntParameter` and removes the no longer necessary custom StorageClass for the `BitVectorAttr` from the SMT dialect that was a workaround for the described issue.

Jezurko committed:
[MLIR][TableGen] Warn on APInt parameter without custom comparator

Jezurko committed:
fixup! [MLIR][TableGen] Warn on APInt parameter without custom comparator

gysit commented on mlir/tools/mlir-tblgen/AttrOrTypeDefGen.cpp:
```diff
@@ -678,8 +678,18 @@ void DefGen::emitStorageClass() {
+    if (param.getCppType().contains("APInt") && !param.hasCustomComparator()) {
+      PrintWarning(
+          def.getLoc(),
+          "Using a raw APInt parameter without a custom comparator is "
+          "discouraged because an assert in the equality operator is "
+          "triggered when the two APInts have different bit widths. This can "
+          "lead to unexpected crashes. Consider using an `APIntParameter` or "
+          "providing a custom comparator.");
```
```suggestion
          "Using a raw APInt parameter without a custom comparator is "
          "not supported because an assert in the equality operator is "
          "triggered when the two APInts have different bit widths. This can "
          "lead to unexpected crashes. Use an `APIntParameter` or "
          "provide a custom comparator.");
```
nit: I would probably strengthen the warning a bit?

gysit reviewed (approved):
Thanks for the follow up!

LGTM but let's give other reviewers a chance to have a look.

Jezurko committed:
fixup!  [MLIR][TableGen] Warn on APInt parameter without custom comparator

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

River707 commented:
The change looks reasonable to me, though I would potentially ask what's the cost of just making this an error? Fine with landing this as-is though.

Jezurko commented:
It could technically break downstream builds that can not experience the issue. For example, even the old version of the Builtin_IntegerAttribute could not trigger the assert, as the type parameter of the attribute contains the bit width information and short-circuited the equality operator.

With that being said, I'm not opposed to making it an error. Fixing the issue is trivial.

Jezurko committed:
Merge branch 'main' into mlir-apint-parameter

xlauko commented:
### Merge activity

* **Apr 22, 2:44 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/135970).

gysit commented:
> It could technically break downstream builds that can not experience the issue. For example, even the old version of the Builtin_IntegerAttribute could not trigger the assert, as the type parameter of the attribute contains the bit width information and short-circuited the equality operator.

True, it sounds like a very doable integration task for downstream projects though? I would thus tend to make it an error since warnings can go unnoticed quite easily. No strong opinion from my side.

Jezurko committed:
fixup! fixup!  [MLIR][TableGen] Warn on APInt parameter without custom comparator

xlauko commented:
LGTM

gysit commented:
PR title and commit message probably should be updated before landing. Otherwise, still LGTM.

Jezurko committed:
fixup! [MLIR][TableGen] Warn on APInt parameter without custom comparator

Jezurko commented:
I've changed the error emitting function to  `printFatalError` since that seems to be the most commonly used error function in tablegen code.

I've updated the PR title and message to match the change. If it's okay like this, can someone merge it for me? Thanks!
